🔥(dc) remove deployment_stamp label for non blue-green deployments

lunika opened the pull request:
## Purpose

Deploying a "non-blue-green application" with the deployment_stamp label in its deployment configuration leads to errors as the deployment_stamp changes at every deployment (OpenShift cannot patch an object that does not exist!).

Fixes #148, #121 

## Proposal

- [x] remove the `deployment_stamp` label of the `mailcatcher` app deployment configuration
- [x] remove the `deployment_stamp` label of the `redis` app deployment configuration

## Update

Removing `label_stamp` is not as easy. Removing it is not enough. Every deploy is based on this stamp and the request made on OKD to find a resource is based on it. So if you just remove the stamp nothing can be deployed.

To resolve this problem and not be dependant on this stamp parameter, we have to know if an application is "stampable".
To do that we will introduce a new resource on each application, a `settings` file containing information describing the application. `databases.yml` should be merged in this new file.
To distinguish "stampable" and not "stampable" objects we will use `is_blue_green_compatible` boolean and depending on this setting we will adjust the request on OKD.

To add this settings we will modify the `apps lookup` and attach on the app object these settings

New things TODO:

- [x] create a `settings.yml` file in `apps/{app}/vars/`.
- [x] parse it in `apps` lookup and inject its content in the `app` variable.
- [x] change request on OKD to reflect this change
- [x] allow deploy of non `blue/green` application

lunika commented:
The description is updated containing information of what we have to do know.

jmaupetit commented:
I've added a reference to #121 as this work will also fix it.

sampaccoud commented:
s/is_blue_green_compatible/use_blue_green ?

lunika committed:
✨(apps) define settings for an app

Each app can need settings to describe a behavior or requirements like
databases. A `settings.yml` file can be added in `apps/{app}/vars` and
will be parsed and injected in `app.settings` variable

lunika committed:
✨(deploy) remove stamp on non blue/green application

Some apps are not blue/green compatible and should be available right
after their deploy. The stamp is not used in this case and the route
will be always available.

lunika commented:
@jmaupetit @sampaccoud this PR is not in WIP anymore

jmaupetit commented:
`use_blue_green` seems more confusing to me as we might think that an app can use blue/green or not. And this is not what we expected in the first place: this flag must stand for the blue/green compatibility.

sampaccoud commented:
@lunika There is another change we need to make before merging this PR. 
The `deploy` playbook should force us to specify what apps in the project we want to deploy. I always forget to specify it and the deployment fails because redis is not blue/green compatible...

After this PR is merged, the deployment will not fail anymore but it will cause a rolling-update redeploy on redis which would cause a (minor) service interruption...

lunika commented:
> After this PR is merged, the deployment will not fail anymore but it will cause a rolling-update redeploy on redis which would cause a (minor) service interruption...

No it will not while you don't change its deployment config, OKD is smart enough to detect no change and will not create a new deployment.

> The deploy playbook should force us to specify what apps in the project we want to deploy. I always forget to specify it and the deployment fails because redis is not blue/green compatible...

I can force to use `apps_filter` but doing this we will don't have the same default behaviour comparing to the other playbooks.

lunika committed:
✨(deploy) prevent deploy playbook to deploy all defined apps

deploy playbook must be used with apps_filter variable to prevent it to
deploy every app defined.

lunika commented:
> The deploy playbook should force us to specify what apps in the project we want to deploy. I always forget to specify it and the deployment fails because redis is not blue/green compatible...

Done

sampaccoud commented:
@lunika sorry I didn't answer earlier. But if pushing all apps does not trigger a redeploy unless the deployment has changed, then my comment is void and we don't need to force the user to add an `apps_filter`.

lunika commented:
route will be replaced because it is in `force` mode so there will be a mini down time

jmaupetit reviewed (commented):
Maybe the `forum` app should also be considered as non-blue-green-compatible? WDYT @sampaccoud ?

sampaccoud commented:
I don't know much about it but I would expect it should be compatible as it relies on Mongodb for persistence...

jmaupetit commented:
Yes it should be compatible, but do we need to deploy it as such as it will almost never get re-deployed? This adds useless complexity to its deployment IMO.

sampaccoud commented:
Then let's keep it blue/green but target deployments per app by default like I proposed initially?

jmaupetit commented on .circleci/config.yml:
```diff
@@ -286,7 +286,7 @@ jobs:
           command: |
             bin/ci-bootstrap "marsha"
-            bin/ci-test-service "marsha" "Api Root" "/api/"
+            bin/ci-test-blue-green-service "marsha" "Api Root" "/api/"
```
I would have preferred that you update the `bin/ci-test-service` script usage to add the blue-green option instead of creating an almost identical script.

lunika replied:
:+1:

lunika commented on bin/ci-test-service:
```diff
@@ -15,10 +15,26 @@ content="${2:-Hello OpenShift! by Arnold}"
+
+# passing an empty string as prefix argument does not work, the default value (next) will
+# be assign to prefix. But we can detect how many arguments was used.
+# if exactly 4 arguments are used and prefix as the default value then we consider that we want to use an empty string
```
s/as/has
